src/main/java/se/liu/ida/hefquin/engine/queryplan/utils/LogicalOpUtils.java

```diff
@@ -2,31 +2,89 @@
 import se.liu.ida.hefquin.engine.federation.FederationMember;
 import se.liu.ida.hefquin.engine.federation.access.BGPRequest;
 import se.liu.ida.hefquin.engine.federation.access.DataRetrievalRequest;
 import se.liu.ida.hefquin.engine.federation.access.SPARQLRequest;
 import se.liu.ida.hefquin.engine.federation.access.TriplePatternRequest;
+import se.liu.ida.hefquin.engine.federation.access.impl.req.SPARQLRequestImpl;
 import se.liu.ida.hefquin.engine.query.BGP;
 import se.liu.ida.hefquin.engine.query.SPARQLGraphPattern;
+import se.liu.ida.hefquin.engine.query.SPARQLGroupPattern;
 import se.liu.ida.hefquin.engine.query.SPARQLQuery;
+import se.liu.ida.hefquin.engine.query.SPARQLUnionPattern;
 import se.liu.ida.hefquin.engine.query.TriplePattern;
 import se.liu.ida.hefquin.engine.query.impl.BGPImpl;
 import se.liu.ida.hefquin.engine.query.impl.GenericSPARQLGraphPatternImpl1;
 import se.liu.ida.hefquin.engine.queryplan.logical.LogicalOperator;
+import se.liu.ida.hefquin.engine.queryplan.logical.LogicalPlan;
 import se.liu.ida.hefquin.engine.queryplan.logical.UnaryLogicalOp;
 import se.liu.ida.hefquin.engine.queryplan.logical.impl.LogicalOpBGPAdd;
 import se.liu.ida.hefquin.engine.queryplan.logical.impl.LogicalOpBGPOptAdd;
 import se.liu.ida.hefquin.engine.queryplan.logical.impl.LogicalOpRequest;
 import se.liu.ida.hefquin.engine.queryplan.logical.impl.LogicalOpTPAdd;
 import se.liu.ida.hefquin.engine.queryplan.logical.impl.LogicalOpTPOptAdd;
+import se.liu.ida.hefquin.engine.queryplan.logical.impl.LogicalPlanWithNullaryRootImpl;
 import se.liu.ida.hefquin.engine.queryplan.physical.PhysicalOperator;
 import se.liu.ida.hefquin.engine.queryplan.physical.PhysicalOperatorForLogicalOperator;
 
 import java.util.HashSet;
 import java.util.Set;
 
 public class LogicalOpUtils {
+	
+	public static LogicalPlan RW(final LogicalOpRequest<?, ?> req) {
+		final FederationMember fm = req.getFederationMember();
+		final VocabularyMapping vm = fm.getVocabularyMapping();
+		
+		if(!(req.getRequest() instanceof TriplePatternRequest)) {
+			// If it's not a TP, return as-is or throw error?
+		}
+		final TriplePatternRequest tpreq = (TriplePatternRequest) req.getRequest(); // Cast the request to be a TriplePatternRequest.
+		final TriplePattern tp = tpreq.getQueryPattern(); // Get the tp.
+		
+		// ApplyVocabularyMapping(tp,vm)
+		final SPARQLGraphPattern newP = vm.translateTriplePattern(tp);
+		
+		return rewriteReqOf(newP, fm);
+		
+	}
+	
+	public static LogicalPlan rewriteReqOf(final SPARQLGraphPattern P, final FederationMember fm) {
+		if (fm.getInterface().supportsBGPRequests()) { // Are all interfaces which support BGP requests SPARQL endpoints? Based on the assumption that there are two types of interfaces: TPF-server and SPARQL-endpoint, and TPF-servers do not.
+			final SPARQLRequest reqP = new SPARQLRequestImpl(P);
+			final LogicalOpRequest<SPARQLRequest, FederationMember> req = new LogicalOpRequest<SPARQLRequest, FederationMember>(fm,reqP);
```

Comment:
If you are checking for support of BGP requests, you then should not create a SPARQL request but, instead, a BGP request (see the `BGPRequest` interface and the implementing class, `BGPRequestImpl`).

However, thinking about this again now, as you know, the result of the `translateTriplePattern` function may not be only a triple pattern or a BGP, but also some other type of graph pattern. Consequently, what you are doing here inside the if-block is correct, but the if-condition is wrong. The condition should actually _**not**_ be whether the interface supports BGP requests but whether the interface is a SPARQL endpoint (because only this type of interface can handle arbitrary graph patterns).